src/StateMachine.php

```diff
@@ -0,0 +1,364 @@
+<?php
+/**
+ * This file is part of the php-state project.
+ *
+ * (c) Yannick Voyer <star.yvoyer@gmail.com> (http://github.com/yvoyer)
+ */
+
+namespace Star\Component\State;
+
+use Star\Component\State\Attribute\StateAttribute;
+use Star\Component\State\Attribute\StringAttribute;
+use Star\Component\State\Builder\TransitionBuilder;
+use Star\Component\State\Event\ContextTransitionWasRequested;
+use Star\Component\State\Event\ContextTransitionWasSuccessful;
+use Star\Component\State\Event\StateEventStore;
+use Star\Component\State\Event\TransitionWasSuccessful;
+use Star\Component\State\Event\TransitionWasRequested;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+
+final class StateMachine
+{
+    /**
+     * @var array
+     *
+     * $whitelist = [
+     *     'state1' => [
+     *         'state2' => allowed,
+     *         'state3' => allowed,
+     *     ],
+     *     'state2' => [
+     *         'state1 => allowed,
+     *     ]
+     * ];
+     */
+    private $whitelist = [];
+
+    /**
+     * @var StateAttribute[]
+     */
+    private $attributes = [];
+
+    /**
+     * @var State[]
+     */
+    private $states = [];
+
+    /**
+     * @var StateTransition[]
+     */
+    private $transitions = [];
+
+    /**
+     * @var EventDispatcher
+     */
+    private $dispatcher;
+
+    /**
+     * @var StateContext
+     */
+    private $context;
+
+    /**
+     * @var FailureHandler
+     */
+    private $failureHandler;
+
+    /**
+     * @var TransitionBuilder
+     */
+    private $transitionFactory;
+
+    /**
+     * @param StateContext $context
+     */
+    public function __construct(StateContext $context)
+    {
+        $this->dispatcher = new EventDispatcher();
+        $this->context = $context;
+        $this->failureHandler = new AlwaysThrowException();
+        $this->transitionFactory = new TransitionBuilder();
+    }
+
+    /**
+     * @param StateContext $context
+     * @param string $transitionName
+     */
+    public function transitContext(StateContext $context, $transitionName)
+    {
+        $from = $context->getCurrentState();
+        $transition = $this->getTransition($transitionName);
+
+        if (! $transition->hasChanged($from)) {
+            return; // no changes detected, do not trigger transition
+        }
+
+        if ($transition->isAllowed($context)) {
```

Comment:
A case where transition would it be allowed on context would be the context do not implement interface